### PR TITLE
Add `hazelcast::new_client()` and `hazelcast::new_client(client_config)` to create new clients - Make `hazelcast_client` constructor private

### PR DESCRIPTION
**Pull request by ihsandemir**
Added the free functions `hazelcast::new_client()` and `hazelcast::new_client(client_config)` to create new clients without using the constructor. The `hazelcast_client` constructor is made as private so that the users will be forced to use the free factory functions to create and start new clients.

### CODE REVIEW

**Comment by ihsandemir** [comment]
verify-linux

**Review by sancar — APPROVED** [review]

**Comment by ihsandemir** [comment]
verify-linux

File: README.md [hunk 1]
```diff
@@ -436,8 +436,8 @@ pass this object to the client when starting it, as shown below.
 ```C++
     hazelcast::client::client_config config;
     config.set_cluster_name("my-cluster"); // the server is configured to use the `my_cluster` as the cluster name hence we need to match it to be able to connect to the server.
-    config.get_network_config().add_address(address("192.168.1.10", 5701));    
-    hazelcast::client::hazelcast_client hz(std::move(config)); // Connects to the cluster member at ip address `192.168.1.10` and port 5701
+config.get_network_config().add_address(address("192.168.1.10", 5701));
```
Comment (yemreinci) [inline README.md:439]:
```suggestion
    config.get_network_config().add_address(address("192.168.1.10", 5701));
```

File: README.md [hunk 2]
```diff
@@ -436,8 +436,8 @@ pass this object to the client when starting it, as shown below.
...
+auto hz = hazelcast::new_client(std::move(config)).get(); // Connects to the cluster member at ip address `192.168.1.10` and port 5701
```
Comment (yemreinci) [inline README.md:440]:
```suggestion
    auto hz = hazelcast::new_client(std::move(config)).get(); // Connects to the cluster member at ip address `192.168.1.10` and port 5701
```

File: README.md [hunk 3]
```diff
@@ -502,8 +502,8 @@ The following example first creates a programmatic configuration object. Then, i
 ```C++
 #include <hazelcast/client/hazelcast_client.h>
 int main() {
-    hazelcast::client::hazelcast_client hz; // Connects to the cluster
-    std::cout << "Started the Hazelcast C++ client instance " << hz.get_name() << std::endl; // Prints client instance name
+auto hz = hazelcast::new_client().get(); // Connects to the cluster
+std::cout << "Started the Hazelcast C++ client instance " << hz.get_name() << std::endl; // Prints client instance name
```
Comment (yemreinci) [inline README.md:506]:
```suggestion
    auto hz = hazelcast::new_client().get(); // Connects to the cluster
    std::cout << "Started the Hazelcast C++ client instance " << hz.get_name() << std::endl; // Prints client instance name
```

File: README.md [hunk 4]
```diff
@@ -545,9 +545,9 @@ Then, you can run the application using the following command:
 ```C++
 #include <hazelcast/client/hazelcast_client.h>
 int main() {
-    hazelcast::client::hazelcast_client hz; // Connects to the cluster
+auto hz = hazelcast::new_client().get(); // Connects to the cluster
```
Comment (yemreinci) [inline README.md:548]:
```suggestion
    auto hz = hazelcast::new_client().get(); // Connects to the cluster
```

File: README.md [hunk 5]
```diff
@@ -545,9 +545,9 @@ Then, you can run the application using the following command:
...
 
-    auto personnel = hz.get_map("personnel_map").get();
+auto personnel = hz.get_map("personnel_map").get();
```
Comment (yemreinci) [inline README.md:550]:
```suggestion
    auto personnel = hz.get_map("personnel_map").get();
```

File: README.md [hunk 6]
```diff
@@ -1568,9 +1568,9 @@ hazelcast::client::topic::reliable_listener make_listener(std::atomic<int> &n_re
 }
 
 void listen_with_default_config() {
-    hazelcast::client::hazelcast_client client;
+hazelcast::client::auto client = hazelcast::new_client().get();
```
Comment (yemreinci) [inline README.md:1571]:
```suggestion
    auto client = hazelcast::new_client().get();
```

**Review by yemreinci — CHANGES_REQUESTED** [review]
I've written suggestions for some of them but there are still many more indentation mistakes in the README.

**Comment by yemreinci** [comment]
A question: Now that the client creation is also asynchronous, would it ever make sense to set `client_connection_strategy_config::set_async_start(true)`?

File: hazelcast/include/hazelcast/client/hazelcast_client.h [hunk 7]
```diff
@@ -25,6 +25,18 @@
 #endif
 
 namespace hazelcast {
+    /**
+    * Constructs a hazelcastClient with default configurations.
+    */
+    boost::future<client::hazelcast_client> HAZELCAST_API new_client();
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/hazelcast_client.h:31]: added declaration above.

File: hazelcast/include/hazelcast/client/hazelcast_client.h [hunk 8]
```diff
@@ -25,6 +25,18 @@
...
+
+    /**
+    * Constructs a hazelcast_client with given config.
+    * Note: client_config will be moved.
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/hazelcast_client.h:35]: Removed the line. It was unnecessary since the API is explaining all it is needed.

**Comment by ihsandemir** [comment]
> A question: Now that the client creation is also asynchronous, would it ever make sense to do `client_connection_strategy_config::set_async_start(true)`?

There is a slight difference, in start_async mode, the `new_client(std::move(config)).get()` may return without any connection to cluster(if any invocation during the time it connects e.g. get_map, it will throw client_offline_exception), while in the other mode it will wait to connect to the cluster. This will work similar also during reconnections. Hence, there is a slight difference. May not be a very useful feature but this is a PM (@burakcelebi ) decision to remove it or not i guess.

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-linux

**Review by yemreinci — APPROVED** [review]

**Pull request merged**